### PR TITLE
Truffle example

### PR DESCRIPTION
**Pull request by 0x3bfc**

### CODE REVIEW

File: content/develop/start/truffle.md [hunk 1]
```diff
@@ -1,5 +1,156 @@
...
+      from: '0x6A33382de9f73B846878a57500d055B981229ac4' // CHANGE THIS ADDRESS
+    },
+```
+
+## Deploying The Contract
+
+To deploy the `CovidVaccineToken` contract, you can run the yarn command as follows:
+
+```
+yarn deploy:aurora
```
Comment (birchmd) [inline content/develop/start/truffle.md:56]: I got an error at this point

```
> erc20-example@1.0.0 compile /mnt/birchmd/tmp/aurora-examples/truffle/erc721-example
> truffle compile

sh: 1: truffle: not found
npm ERR! code ELIFECYCLE
npm ERR! syscall spawn
npm ERR! file sh
npm ERR! errno ENOENT
npm ERR! erc20-example@1.0.0 compile: `truffle compile`
npm ERR! spawn ENOENT
npm ERR!
npm ERR! Failed at the erc20-example@1.0.0 compile script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/birchmd/.npm/_logs/2021-05-27T13_58_34_258Z-debug.log
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Seems like there is some missing step where `truffle` was supposed to be installed. I thought calling `yarn` above was supposed to install all the dependencies?

**Review by birchmd — CHANGES_REQUESTED** [review]
Most comments are just small grammatical fixes, but the blocker for me is the error I got trying to deploy. Some assistance here would be appreciated, and if necessary we can add more details to the tutorial to prevent others from encountering the same error.

File: content/develop/start/truffle.md [hunk 2]
```diff
@@ -1,5 +1,156 @@
...
+4. Now the `minter` can redistribute that ticket 🎫 to new participant in the line.
+5. And now that participant have access to the same ticket that have been spended by participant-1.
+
+
+## Installing Prerequisites
+This tutorial assumes that you have `Node 12+` and `yarn`. Please refer to the [Yarn installation page](https://classic.yarnpkg.com/en/docs/install) if you don't have `yarn` installed locally.
+
+To install the prerequisites packages, clone the examples code then run `yarn` in order to install the prerequisites:
+```
+git clone https://github.com/aurora-is-near/aurora-examples.git
```
[the remark this answers is not in the document]
Comment (0x3bfc) [reply content/develop/start/truffle.md:31]: Yes, it should go public once we merge this PR.

File: content/develop/start/truffle.md [hunk 3]
```diff
@@ -1,5 +1,156 @@
...
+This example is originally forked from [OpenZeppelin examples](https://docs.openzeppelin.com/contracts/4.x/erc721). However it implements a simple COVID-19 vaccine token 💊💊. This tutorial assumes that you are familiar with non-fungible tokens (NFT) concept. For more details about the non-fungible token standard (ERC721), please refer to [OpenZeppelin ERC721 implementation](https://github.com/OpenZeppelin/openzeppelin-contracts/tree/1b37c21da58f6379cfe09c0140cf56d67b19a0bc/contracts/token/ERC721).
+
+## COVID Vaccine NFT
+
+![](../../_img/Truffle-nft-aurora-example.png)
+
+1. The `minter` address (which is managed by the vaccination program manager) can distribute (mint) the vaccine tickets (NFT tokens 💊💊💊) to the people who are part of the vaccination program. 
+2. When a participant receives the token 💊, he/she can get access to the vaccine by spending the NFT token. 
+3. This means whether burning the NFT token or sending it back to the minter address.
+4. Now the `minter` can redistribute that ticket 🎫 to new participant in the line.
```
[the remark this answers is not in the document]
Comment (0x3bfc) [reply content/develop/start/truffle.md:22]: I updated this part in the last commit.

File: content/develop/start/truffle.md [hunk 4]
```diff
@@ -1,5 +1,156 @@
...
+yarn 
+```
+
+## Connecting Truffle to Aurora
+
+Export your `MNEMONIC` as follows:
+```
+export MNEMONIC='YOUR MNEMONICS HERE'
+```
+The `truffle-config.js` will pickup your `MNEMONIC` and extract the address that will be used for sending, and signing transactions on Aurora network.
```
[the remark this answers is not in the document]
Comment (0x3bfc) [reply content/develop/start/truffle.md:42]: Yes, you are right ... will update it.

File: content/develop/start/truffle.md (hunk 1 above)
Comment (0x3bfc) [reply content/develop/start/truffle.md:61]: Will update the doc

File: content/develop/start/truffle.md [hunk 5]
```diff
@@ -1,5 +1,180 @@
...
+You should change the `from` back to the minter address before continue to the next step.
+### Transfer
+
+Participant can transfer the token to the `minter` after receiving the vaccine. As shown below a participant can only send the NFT token if the receiver for this token is the minter (`0x6A33382de9f73B846878a57500d055B981229ac4`). 
+
+```bash
+% truffle console --network aurora
+truffle(aurora)> const cvt = await CovidVaccineToken.deployed()
+truffle(aurora)> await cvt.ownerOf(1)
+truffle(aurora)> await cvt.safeTransferFrom('0x2531a4D108619a20ACeE88C4354a50e9aC48ecfe', '0x6A33382de9f73B846878a57500d055B981229ac4', 1)
```
Comment (birchmd) [inline content/develop/start/truffle.md:162]: I'm stuck here now. I get the following Error:

```
Uncaught Error: execution reverted:
    at evalmachine.<anonymous>:1:13
    at evalmachine.<anonymous>:2:49
    at sigintHandlersWrap (vm.js:272:15)
    at Script.runInContext (vm.js:127:14)
    at runScript (/mnt/birchmd/tmp/aurora-examples/truffle/erc721-example/node_modules/truffle/build/webpack:/packages/core/lib/console.js:270:1)
    at Console.interpret (/mnt/birchmd/tmp/aurora-examples/truffle/erc721-example/node_modules/truffle/build/webpack:/packages/core/lib/console.js:285:1)
    at bound (domain.js:427:14)
    at REPLServer.runBound [as eval] (domain.js:440:12)
    at REPLServer.onLine (repl.js:777:10)
    at REPLServer.emit (events.js:314:20)
    at REPLServer.EventEmitter.emit (domain.js:483:12)
    at REPLServer.Interface._onLine (readline.js:329:10)
    at REPLServer.Interface._line (readline.js:658:8)
    at REPLServer.Interface._ttyWrite (readline.js:1003:14)
    at REPLServer.self._ttyWrite (repl.js:867:9)
    at ReadStream.onkeypress (readline.js:205:10)
    at ReadStream.emit (events.js:314:20)
    at ReadStream.EventEmitter.emit (domain.js:483:12)
    at emitKeys (internal/readline/utils.js:335:14)
    at emitKeys.next (<anonymous>) {
  reason: 'ERC721: transfer caller is not owner nor approved',
  hijackedStack: 'Error: execution reverted:\n' +
    '    at /mnt/birchmd/tmp/aurora-examples/truffle/erc721-example/node_modules/truffle-hdwallet-provider/dist/webpack:/truffle-hdwallet-provider/Users/gnidan/src/work/truffle/node_modules/web3-provider-engine/subproviders/provider.js:18:1\n' +
    '    at t.callback [as onreadystatechange] (/mnt/birchmd/tmp/aurora-examples/truffle/erc721-example/node_modules/truffle-hdwallet-provider/dist/webpack:/truffle-hdwallet-provider/Users/gnidan/src/work/truffle/node_modules/web3/node_modules/web3-providers-http/src/index.js:96:1)\n' +
    '    at t._a [as dispatchEvent] (/mnt/birchmd/tmp/aurora-examples/truffle/erc721-example/node_modules/truffle-hdwallet-provider/dist/webpack:/truffle-hdwallet-provider/Users/gnidan/src/work/truffle/node_modules/xhr2-cookies/dist/xml-http-request-event-target.js:27:61)\n' +
    '    at t.dispatchEvent [as _setReadyState] (/mnt/birchmd/tmp/aurora-examples/truffle/erc721-example/node_modules/truffle-hdwallet-provider/dist/webpack:/truffle-hdwallet-provider/Users/gnidan/src/work/truffle/node_modules/xhr2-cookies/dist/xml-http-request.js:208:1)\n' +
    '    at t._setReadyState [as _onHttpResponseEnd] (/mnt/birchmd/tmp/aurora-examples/truffle/erc721-example/node_modules/truffle-hdwallet-provider/dist/webpack:/truffle-hdwallet-provider/Users/gnidan/src/work/truffle/node_modules/xhr2-cookies/dist/xml-http-request.js:318:1)\n' +
    '    at IncomingMessage._onHttpResponseEnd (/mnt/birchmd/tmp/aurora-examples/truffle/erc721-example/node_modules/truffle-hdwallet-provider/dist/webpack:/truffle-hdwallet-provider/Users/gnidan/src/work/truffle/node_modules/xhr2-cookies/dist/xml-http-request.js:289:47)\n' +
    '    at IncomingMessage.emit (events.js:326:22)\n' +
    '    at IncomingMessage.EventEmitter.emit (domain.js:506:15)\n' +
    '    at endReadableNT (_stream_readable.js:1241:12)\n' +
    '    at processTicksAndRejections (internal/process/task_queues.js:84:21)'
}
```

I did change the `truffle-config.js` back to the minter as instructed. I'm not sure what I am doing wrong.

**Review by birchmd — CHANGES_REQUESTED** [review]
Thanks for taking care of my previous comments! I got farther this time, but still not quite to the end. Further help is appreciated.

File: content/develop/start/truffle.md (hunk 5 above)
Comment (0x3bfc) [reply content/develop/start/truffle.md]: I've resolved this in the examples repo. Please make pull and try it again.

File: content/develop/start/truffle.md [hunk 6]
```diff
@@ -1,5 +1,180 @@
...
+
+```bash
+% truffle console --network aurora
+truffle(aurora)> const cvt = await CovidVaccineToken.deployed()
+truffle(aurora)> await cvt.safeTransferFrom('0x2531a4d108619a20acee88c4354a50e9ac48ecfe', '0x8722C88e82AbCC639148Ab6128Cd63333B2Ad771', 1) 
+Uncaught:
+Error: Unknown address - unable to sign transaction for this address: "0x2531a4d108619a20acee88c4354a50e9ac48ecfe"
+...
+reason: 'Invalid Transfer',
+  hijackedStack: 'Error: Unknown address - unable to sign transaction for this address: "0x2531a4d108619a20acee88c4354a50e9ac48ecfe"\n'
```
[the remark this answers is not in the document]
Comment (0x3bfc) [reply content/develop/start/truffle.md]: Yes, this was fixed in the examples repo :)

File: content/develop/start/truffle.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (artob) [reply content/develop/start/truffle.md:63]: Done

**Review by artob — APPROVED** [review]
LGTM :+1: An interesting use case!

@0x3bfc For future documentation writing, please note the fixes I made in 1cb08a78f6909cb5e4c309ad03b16e4749b77492 and 0bae84138431c987138609f4033189641f18248e. Also, hard wrap at 80 columns.

@birchmd Thanks for your thorough review and testing!

File: content/develop/start/truffle.md [hunk 7]
```diff
@@ -1,5 +1,224 @@
...
+```
+
+### Burn tokens
+
+This is an alternative scenario for the NFT token lifecycle. Instead of
+transfering the token back to the minter, the participant can decide to burn the
+NFT token by calling the `burn` function:
+
+```bash
+truffle(aurora)> await cvt.burn(1, {from: accounts[1]}) // 1 is the tokenID
```
Comment (birchmd) [inline content/develop/start/truffle.md:207]: Is it a bug that a non-owner of the token can burn? When I was going through the tutorial I was able to burn the token with `accounts[1]` even after I had transferred ownership back to `accounts[0]`.

**Pull request merged**